src/Sulu/Bundle/AdminBundle/Metadata/Form/FormMetadataProvider.php

```diff
@@ -0,0 +1,386 @@
+        StructureMetadataFactory $structureMetadataFactory,
+        ExpressionLanguage $expressionLanguage,
+        array $locales,
+        array $formDirectories,
+        string $cacheDir,
+        bool $debug
+    ) {
+        $this->formXmlLoader = $formXmlLoader;
+        $this->structureMetadataFactory = $structureMetadataFactory;
+        $this->expressionLanguage = $expressionLanguage;
+        $this->locales = $locales;
+        $this->formDirectories = $formDirectories;
+        $this->cacheDir = $cacheDir;
+        $this->debug = $debug;
+    }
+
+    public function getMetadata(string $key, string $locale)
+    {
+        $configCache = $this->getConfigCache($key, $locale);
+
+        if (!$configCache->isFresh()) {
+            $this->warmUp($this->cacheDir);
+        }
+
+        if (!file_exists($configCache->getPath())) {
+            throw new MetadataNotFoundException('form', $key);
+        }
+
+        $form = unserialize(file_get_contents($configCache->getPath()));
+
+        if ($form instanceof Form) {
+            $this->evaluateFormItemExpressions($form->getItems());
+        } elseif ($form instanceof TypedForm) {
+            foreach ($form->getForms() as $formType) {
+                $this->evaluateFormItemExpressions($formType->getItems());
+            }
+        }
+
+
+        return $form;
+    }
+
+    /**
+     * @param Item[] $items
+     */
+    private function evaluateFormItemExpressions(array $items)
+    {
+        foreach ($items as $item) {
+            if ($item instanceof Section) {
+                $this->evaluateFormItemExpressions($item->getItems());
+            }
+
+            if ($item instanceof Field) {
+                foreach ($item->getTypes() as $type) {
+                    $this->evaluateFormItemExpressions($type->getForm()->getItems());
+                }
+
+                foreach ($item->getOptions() as $option) {
+                    if (Option::TYPE_EXPRESSION === $option->getType()) {
+                        $option->setValue($this->expressionLanguage->evaluate($option->getValue()));
+                    }
+                }
+            }
+        }
+    }
+
+    public function warmUp($cacheDir)
+    {
+        $formsMetadata = [];
+        $formFinder = (new Finder())->in($this->formDirectories)->name('*.xml');
+        foreach ($formFinder as $formFile) {
+            $formMetadata = $this->formXmlLoader->load($formFile->getPathName());
+            $formsMetadata[$formMetadata->getKey()] = $formMetadata;
+        }
+
+        $structuresMetadataByTypes = [];
+        foreach ($this->structureMetadataFactory->getStructureTypes() as $structureType) {
+            foreach ($this->structureMetadataFactory->getStructures($structureType) as $structureMetadata) {
+                if ($structureMetadata->isInternal()) {
+                    continue;
+                }
+
+                $structuresMetadataByTypes[$structureType][] = $structureMetadata;
```

Comment:
The `structureType` contains the values `home`, `page` and `snippet` from the DocumentManager in the default installation. 
There are two problems with that:

- The values are written in singular, whereby the keys in the frontend have been plural so far. I could use some Inflector service, or we decide for one or the other. Whereby I would probably feel less comfortable with changing the names in the DocumentManager...
- `page` and `home` are basically the same, and we either have to skip one of them here (question is how) or just ignore the fact that both are generated. Shouldn't be a too big problem in practice, because the metadata for `home` will never be called anyway.

Reply:
- skip home
- use inflektor instead of renaming type